File: lib/forecasting.ts
export interface UsagePattern {
  itemId: string;
  itemName: string;
  dailyUsage: number[];
  weeklyAverage: number;
  trend: "increasing" | "decreasing" | "stable";
  seasonality: number;
  confidence: number;
}

export interface ForecastData {
  itemId: string;
  itemName: string;
  currentStock: number;
  predictedUsage: number[];
  daysUntilStockout: number;
  recommendedReorderPoint: number;
  recommendedOrderQuantity: number;
  confidence: number;
  riskLevel: "low" | "medium" | "high";
}

export interface AnalyticsData {
  totalSpend: number;
  monthlySpend: number[];
  topExpensiveItems: Array<{
    name: string;
    totalCost: number;
    usage: number;
  }>;
  categoryBreakdown: Array<{
    category: string;
    items: number;
    totalValue: number;
    percentage: number;
  }>;
  usageTrends: Array<{
    date: string;
    totalUsage: number;
    categories: Record<string, number>;
  }>;
}

export interface SimulationResult {
  scenario: string;
  impactedItems: Array<{
    itemName: string;
    currentStock: number;
    projectedStock: number;
    daysUntilStockout: number;
    additionalOrderNeeded: number;
  }>;
  totalAdditionalCost: number;
  criticalItems: string[];
}

class SimpleNeuralNetwork {
  private weights: number[];
  private bias: number;

  constructor(inputSize: number) {
    this.weights = Array(inputSize)
      .fill(0)
      .map(() => Math.random() * 0.1 - 0.05);
    this.bias = Math.random() * 0.1 - 0.05;
  }

  predict(inputs: number[]): number {
    const sum =
      inputs.reduce((acc, input, i) => acc + input * this.weights[i], 0) +
      this.bias;
    return Math.max(0, sum);
  }

  train(inputs: number[], target: number, learningRate: number = 0.01): void {
    const prediction = this.predict(inputs);
    const error = target - prediction;

    this.weights = this.weights.map(
      (weight, i) => weight + learningRate * error * inputs[i]
    );
    this.bias += learningRate * error;
  }
}

const calculateTrend = (
  data: number[]
): { slope: number; trend: "increasing" | "decreasing" | "stable" } => {
  const n = data.length;
  const x = Array.from({ length: n }, (_, i) => i);
  const y = data;

  const sumX = x.reduce((a, b) => a + b, 0);
  const sumY = y.reduce((a, b) => a + b, 0);
  const sumXY = x.reduce((sum, xi, i) => sum + xi * y[i], 0);
  const sumXX = x.reduce((sum, xi) => sum + xi * xi, 0);

  const slope = (n * sumXY - sumX * sumY) / (n * sumXX - sumX * sumX);

  let trend: "increasing" | "decreasing" | "stable";
  if (Math.abs(slope) < 0.1) trend = "stable";
  else if (slope > 0) trend = "increasing";
  else trend = "decreasing";

  return { slope, trend };
};

const movingAverage = (data: number[], window: number = 7): number[] => {
  const result = [];
  for (let i = 0; i < data.length; i++) {
    const start = Math.max(0, i - window + 1);
    const slice = data.slice(start, i + 1);
    result.push(slice.reduce((a, b) => a + b, 0) / slice.length);
  }
  return result;
};

export class ForecastingEngine {
  private usageHistory: Map<string, number[]> = new Map();
  private neuralNetworks: Map<string, SimpleNeuralNetwork> = new Map();
  private trained: boolean = false;

  async initializeWithRealData(): Promise<void> {
    if (this.trained) return;

    try {
      const { getInventoryItems } = await import("@/lib/inventory");
      const items = await getInventoryItems();

      for (const item of items) {
        const usageData = await this.fetchHistoricalUsage(item.id);
        if (usageData.length > 0) {
          this.usageHistory.set(item.id, usageData);
          this.trainNeuralNetwork(item.id, usageData);
        } else {
          this.usageHistory.set(
            item.id,
            this.generateBaselineUsage(item.currentStock)
          );
        }
      }

      this.trained = true;
    } catch (error) {
      console.error("Failed to initialize forecasting engine:", error);
    }
  }

  private async fetchHistoricalUsage(itemId: string): Promise<number[]> {
    try {
      const response = await fetch(
        `http://localhost:8000/api/usage-logs/${itemId}`
      );
      if (!response.ok) return [];

      const data = await response.json();
      const logs = data.logs || [];

      const dailyUsage: Record<string, number> = {};
      logs.forEach((log: any) => {
        const date = new Date(log.timestamp).toISOString().split("T")[0];
        dailyUsage[date] = (dailyUsage[date] || 0) + log.quantity;
      });

      const dates = Object.keys(dailyUsage).sort();
      return dates.map((date) => dailyUsage[date]).slice(-30);
    } catch (error) {
      return [];
    }
  }

  private generateBaselineUsage(currentStock: number): number[] {
    const baseUsage = Math.max(1, Math.floor(currentStock / 30));
    return Array(30)
      .fill(0)
      .map(() => Math.floor(baseUsage * (0.5 + Math.random())));
  }

  private trainNeuralNetwork(itemId: string, usageData: number[]): void {
    const nn = new SimpleNeuralNetwork(7);

    for (let i = 7; i < usageData.length; i++) {
      const inputs = usageData.slice(i - 7, i);
      const target = usageData[i];
      nn.train(inputs, target);
    }

    this.neuralNetworks.set(itemId, nn);
  }

  async analyzeUsagePattern(
    itemId: string,
    itemName: string
  ): Promise<UsagePattern> {
    await this.initializeWithRealData();

    const usage = this.usageHistory.get(itemId) || [];
    if (usage.length === 0) {
      return {
        itemId,
        itemName,
        dailyUsage: [],
        weeklyAverage: 0,
        trend: "stable",
        seasonality: 0,
        confidence: 0.5,
      };
    }

    const smoothed = movingAverage(usage, 7);
    const { trend } = calculateTrend(smoothed);
    const weeklyAverage =
      usage.slice(0, 7).reduce((a, b) => a + b, 0) /
      Math.max(1, Math.min(7, usage.length));
    const seasonality = this.calculateSeasonality(usage);

    return {
      itemId,
      itemName,
      dailyUsage: usage,
      weeklyAverage,
      trend,
      seasonality,
      confidence: Math.min(0.95, 0.5 + usage.length / 100),
    };
  }

  private calculateSeasonality(usage: number[]): number {
    if (usage.length < 7) return 0;

    const dayAverages = Array(7).fill(0);
    const dayCounts = Array(7).fill(0);

    usage.forEach((value, index) => {
      const dayOfWeek = index % 7;
      dayAverages[dayOfWeek] += value;
      dayCounts[dayOfWeek]++;
    });

    for (let i = 0; i < 7; i++) {
      dayAverages[i] = dayCounts[i] > 0 ? dayAverages[i] / dayCounts[i] : 0;
    }

    const overallAverage = dayAverages.reduce((a, b) => a + b, 0) / 7;
    if (overallAverage === 0) return 0;

    const variance =
      dayAverages.reduce(
        (sum, avg) => sum + Math.pow(avg - overallAverage, 2),
        0
      ) / 7;
    return Math.sqrt(variance) / overallAverage;
  }

  async generateForecast(
    itemId: string,
    itemName: string,
    currentStock: number,
    minThreshold: number
  ): Promise<ForecastData> {
    const pattern = await this.analyzeUsagePattern(itemId, itemName);
    const forecastDays = 30;

    const predictedUsage = [];
    let trendMultiplier = 1;

    for (let day = 1; day <= forecastDays; day++) {
      let predicted: number;

      const nn = this.neuralNetworks.get(itemId);
      if (nn && pattern.dailyUsage.length >= 7) {
        const recentUsage = pattern.dailyUsage.slice(-7);
        predicted = Math.max(0, Math.round(nn.predict(recentUsage)));
      } else {
        if (pattern.trend === "increasing") trendMultiplier += 0.01;
        else if (pattern.trend === "decreasing") trendMultiplier -= 0.005;

        const dayOfWeek = day % 7;
        const seasonalMultiplier =
          1 + Math.sin((dayOfWeek * Math.PI) / 3.5) * pattern.seasonality;
        predicted = Math.max(
          0,
          Math.round(
            pattern.weeklyAverage * trendMultiplier * seasonalMultiplier
          )
        );
      }

      predictedUsage.push(predicted);
    }

    let remainingStock = currentStock;
    let daysUntilStockout = forecastDays;

    for (let i = 0; i < predictedUsage.length; i++) {
      remainingStock -= predictedUsage[i];
      if (remainingStock <= 0) {
        daysUntilStockout = i + 1;
        break;
      }
    }

    const avgDailyUsage = pattern.weeklyAverage;
    const leadTimeDays = 7;
    const safetyStock = avgDailyUsage * 3;

    const recommendedReorderPoint = avgDailyUsage * leadTimeDays + safetyStock;
    const recommendedOrderQuantity = Math.max(
      minThreshold * 2,
      avgDailyUsage * 30
    );

    let riskLevel: "low" | "medium" | "high";
    if (daysUntilStockout <= 3) riskLevel = "high";
    else if (daysUntilStockout <= 7) riskLevel = "medium";
    else riskLevel = "low";

    return {
      itemId,
      itemName,
      currentStock,
      predictedUsage,
      daysUntilStockout,
      recommendedReorderPoint,
      recommendedOrderQuantity,
      confidence: pattern.confidence,
      riskLevel,
    };
  }

  async generateAnalytics(items: any[]): Promise<AnalyticsData> {
    const totalSpend = items.reduce(
      (sum, item) => sum + item.currentStock * 10,
      0
    );

    const monthlySpend = Array.from({ length: 12 }, (_, i) => {
      const baseSpend = totalSpend * 0.8;
      const seasonal = Math.sin(((i + 1) * Math.PI) / 6) * totalSpend * 0.2;
      return Math.max(
        0,
        baseSpend + seasonal + (Math.random() - 0.5) * totalSpend * 0.1
      );
    });

    const topExpensiveItems = items
      .map((item) => ({
        name: item.name,
        totalCost: item.currentStock * 10,
        usage:
          this.usageHistory
            .get(item.id)
            ?.slice(0, 7)
            .reduce((a, b) => a + b, 0) || 0,
      }))
      .sort((a, b) => b.totalCost - a.totalCost)
      .slice(0, 5);

    const categoryMap = new Map<
      string,
      { items: number; totalValue: number }
    >();
    items.forEach((item) => {
      const category = "General";
      const existing = categoryMap.get(category) || { items: 0, totalValue: 0 };
      categoryMap.set(category, {
        items: existing.items + 1,
        totalValue: existing.totalValue + item.currentStock * 10,
      });
    });

    const categoryBreakdown = Array.from(categoryMap.entries()).map(
      ([category, data]) => ({
        category,
        items: data.items,
        totalValue: data.totalValue,
        percentage: (data.totalValue / totalSpend) * 100,
      })
    );

    const usageTrends = Array.from({ length: 30 }, (_, i) => {
      const date = new Date();
      date.setDate(date.getDate() - (29 - i));

      const categories: Record<string, number> = {};
      let totalUsage = 0;

      items.forEach((item) => {
        const usage = this.usageHistory.get(item.id)?.[29 - i] || 0;
        totalUsage += usage;
        categories["General"] = (categories["General"] || 0) + usage;
      });

      return {
        date: date.toISOString().split("T")[0],
        totalUsage,
        categories,
      };
    });

    return {
      totalSpend,
      monthlySpend,
      topExpensiveItems,
      categoryBreakdown,
      usageTrends,
    };
  }

  async runSimulation(
    items: any[],
    scenario: string,
    multiplier: number
  ): Promise<SimulationResult> {
    const impactedItems = await Promise.all(
      items.map(async (item) => {
        const forecast = await this.generateForecast(
          item.id,
          item.name,
          item.currentStock,
          item.minStock
        );
        const adjustedUsage = forecast.predictedUsage.map((usage) =>
          Math.round(usage * multiplier)
        );

        let remainingStock = item.currentStock;
        let daysUntilStockout = 30;

        for (let i = 0; i < adjustedUsage.length; i++) {
          remainingStock -= adjustedUsage[i];
          if (remainingStock <= 0) {
            daysUntilStockout = i + 1;
            break;
          }
        }

        const projectedStock = Math.max(0, remainingStock);
        const additionalOrderNeeded =
          remainingStock < 0 ? Math.abs(remainingStock) : 0;

        return {
          itemName: item.name,
          currentStock: item.currentStock,
          projectedStock,
          daysUntilStockout,
          additionalOrderNeeded,
        };
      })
    );

    const totalAdditionalCost = impactedItems.reduce((sum, item) => {
      return sum + item.additionalOrderNeeded * 10;
    }, 0);

    const criticalItems = impactedItems
      .filter((item) => item.daysUntilStockout <= 7)
      .map((item) => item.itemName);

    return {
      scenario,
      impactedItems,
      totalAdditionalCost,
      criticalItems,
    };
  }

  async addUsageData(itemId: string, usage: number): Promise<void> {
    const existing = this.usageHistory.get(itemId) || [];
    existing.unshift(usage);
    existing.splice(30);
    this.usageHistory.set(itemId, existing);

    const nn = this.neuralNetworks.get(itemId);
    if (nn && existing.length >= 8) {
      const inputs = existing.slice(1, 8);
      nn.train(inputs, usage);
    }
  }
}

let forecastingEngine: ForecastingEngine | null = null;

export const getForecastingEngine = (): ForecastingEngine => {
  if (!forecastingEngine) {
    forecastingEngine = new ForecastingEngine();
  }
  return forecastingEngine;
};

File: server/components/forecasting.py
from __future__ import annotations
from typing import List, Dict, Any
import math
try:
	import numpy as np
	HAS_NUMPY = True
except Exception:
	np = None
	HAS_NUMPY = False


def _safe_number(x, default=0.0):
	try:
		v = float(x)
		if math.isnan(v) or math.isinf(v):
			return default
		return v
	except Exception:
		return default


def _build_daily_usage_series(logs: List[dict], days: int = 30):
	from datetime import datetime, timedelta
	end = datetime.utcnow().date()
	start = end - timedelta(days=days - 1)
	dates = [start + timedelta(days=i) for i in range(days)]
	date_index = {d.isoformat(): i for i, d in enumerate(dates)}
	series: Dict[str, Any] = {}
	for log in logs:
		item_id = log.get("itemId")
		ts = log.get("timestamp")
		qty = int(log.get("quantity", 0))
		if not item_id or ts is None:
			continue
		d = ts.date().isoformat() if hasattr(ts, "date") else str(ts)[:10]
		if d not in date_index:
			continue
		idx = date_index[d]
		if item_id not in series:
			if HAS_NUMPY:
				series[item_id] = (np.zeros(days, dtype=float))
			else:
				series[item_id] = [0.0] * days
		if HAS_NUMPY:
			series[item_id][idx] += max(0, qty)
		else:
			series[item_id][idx] = float(series[item_id][idx]) + max(0, qty)
	return series


def _moving_average(x, window: int = 7):
	if HAS_NUMPY:
		if x.size == 0:
			return x
		window = max(1, min(window, x.size))
		cumsum = np.cumsum(np.insert(x, 0, 0.0))
		ma = (cumsum[window:] - cumsum[:-window]) / window
		pad = np.full(x.size - ma.size, ma[0] if ma.size > 0 else 0.0)
		return np.concatenate([pad, ma])
	else:
		n = len(x)
		if n == 0:
			return []
		window = max(1, min(window, n))
		out = []
		for i in range(n):
			start = max(0, i - window + 1)
			sl = x[start : i + 1]
			avg = sum(sl) / max(1, len(sl))
			out.append(avg)
		return out


def _linear_trend(x) -> float:
	if HAS_NUMPY:
		n = x.size
		if n < 2:
			return 0.0
		t = np.arange(n)
		t_mean = t.mean()
		x_mean = x.mean()
		denom = ((t - t_mean) ** 2).sum()
		if denom == 0:
			return 0.0
		slope = ((t - t_mean) * (x - x_mean)).sum() / denom
		return float(slope)
	else:
		n = len(x)
		if n < 2:
			return 0.0
		t = list(range(n))
		t_mean = sum(t) / n
		x_mean = sum(x) / n
		denom = sum((ti - t_mean) ** 2 for ti in t)
		if denom == 0:
			return 0.0
		num = sum((ti - t_mean) * (xi - x_mean) for ti, xi in zip(t, x))
		return float(num / denom)


def forecast_usage(daily_usage, horizon: int = 30):
	if HAS_NUMPY:
		if daily_usage.size == 0:
			return np.zeros(horizon)
		ma = _moving_average(daily_usage, window=min(7, max(1, daily_usage.size)))
		slope = _linear_trend(ma)
		base = ma[-1] if (hasattr(ma, 'size') and ma.size > 0) else daily_usage.mean()
		t = np.arange(1, horizon + 1)
		pred = np.maximum(0.0, base + slope * t)
		ripple = 1.0 + 0.1 * np.sin(2 * np.pi * (t % 7) / 7)
		return np.maximum(0.0, np.round(pred * ripple))
	else:
		if len(daily_usage) == 0:
			return [0.0] * horizon
		ma = _moving_average(daily_usage, window=min(7, max(1, len(daily_usage))))
		slope = _linear_trend(ma)
		base = ma[-1] if len(ma) > 0 else (sum(daily_usage) / max(1, len(daily_usage)))
		preds = []
		for i in range(1, horizon + 1):
			pred = max(0.0, base + slope * i)
			ripple = 1.0 + 0.1 * math.sin(2 * math.pi * (i % 7) / 7)
			preds.append(max(0.0, round(pred * ripple)))
		return preds


def compute_forecasts_and_analytics(items: List[Any], logs: List[Any]) -> tuple[list[dict], dict]:
	raw_logs = [l.model_dump() if hasattr(l, "model_dump") else dict(l) for l in logs]
	usage_map = _build_daily_usage_series(raw_logs, days=30)
	forecasts: list[dict] = []
	total_spend = 0.0
	top_items: list[dict] = []
	category_values: Dict[str, float] = {}
	for item in items:
		item_id = getattr(item, 'id', None)
		name = getattr(item, 'name', 'Unknown')
		current_stock = int(getattr(item, 'currentStock', 0) or 0)
		min_stock = int(getattr(item, 'minStock', 0) or 0)
		price = _safe_number(getattr(item, 'price', 10.0) or 10.0, 10.0)
		category = getattr(item, 'category', None) or 'General'
		total_spend += current_stock * price
		category_values[category] = category_values.get(category, 0.0) + current_stock * price
		usage = usage_map.get(item_id, (np.zeros(30) if HAS_NUMPY else [0.0] * 30))
		pred = forecast_usage(usage, horizon=30)
		remaining = current_stock
		days_until = 30
		for i, u in enumerate(pred):
			remaining -= int(u)
			if remaining <= 0:
				days_until = i + 1
				break
		if HAS_NUMPY:
			avg_daily = float(usage.mean()) if getattr(usage, 'size', 0) > 0 else 0.0
		else:
			avg_daily = (sum(usage) / max(1, len(usage))) if len(usage) > 0 else 0.0
		lead_time = 7
		safety_stock = avg_daily * 3
		reorder_point = int(round(avg_daily * lead_time + safety_stock))
		order_qty = int(max(min_stock * 2, avg_daily * 30))
		if days_until <= 3:
			risk = 'high'
		elif days_until <= 7:
			risk = 'medium'
		else:
			risk = 'low'
		size_val = (getattr(usage, 'size', None) or len(usage)) if not HAS_NUMPY else usage.size
		confidence = float(min(0.95, 0.5 + (size_val / 100)))
		predicted_usage = pred.tolist() if HAS_NUMPY else pred
		forecasts.append({
			'itemId': item_id,
			'itemName': name,
			'currentStock': current_stock,
			'predictedUsage': predicted_usage,
			'daysUntilStockout': int(days_until),
			'recommendedReorderPoint': int(max(0, reorder_point)),
			'recommendedOrderQuantity': int(max(0, order_qty)),
			'confidence': confidence,
			'riskLevel': risk,
		})
		if HAS_NUMPY:
			recent_usage = int(usage[-7:].sum()) if getattr(usage, 'size', 0) >= 7 else int(usage.sum())
		else:
			recent_usage = int(sum(usage[-7:]))
		top_items.append({
			'name': name,
			'totalCost': max(0.0, current_stock * price),
			'usage': max(0, recent_usage),
		})
	top_items.sort(key=lambda x: x['totalCost'], reverse=True)
	top_items = top_items[:5]
	monthly = []
	for i in range(12):
		base = total_spend * 0.8
		seasonal = (math.sin(((i + 1) * math.pi) / 6.0)) * total_spend * 0.2
		monthly.append(float(max(0.0, base + seasonal)))
	total_value = float(max(0.0, total_spend))
	category_breakdown = []
	for cat, val in category_values.items():
		perc = (val / total_value * 100.0) if total_value > 0 else 0.0
		category_breakdown.append({
			'category': cat,
			'items': sum(1 for it in items if (getattr(it, 'category', None) or 'General') == cat),
			'totalValue': float(max(0.0, val)),
			'percentage': float(max(0.0, perc)),
		})
	from datetime import datetime, timedelta
	end = datetime.utcnow().date()
	start = end - timedelta(days=29)
	dates = [start + timedelta(days=i) for i in range(30)]
	trends = []
	for d in dates:
		d_iso = d.isoformat()
		total = 0.0
		categories: Dict[str, float] = {}
		for item in items:
			cat = getattr(item, 'category', None) or 'General'
			u = usage_map.get(getattr(item, 'id'), (np.zeros(30) if HAS_NUMPY else [0.0] * 30))
			idx = (d - start).days
			if HAS_NUMPY:
				length = int(getattr(u, 'size', 0))
			else:
				length = len(u)
			val = float(u[idx]) if 0 <= idx < length else 0.0
			total += val
			categories[cat] = categories.get(cat, 0.0) + val
		trends.append({
			'date': d_iso,
			'totalUsage': float(max(0.0, total)),
			'categories': categories,
		})
	analytics = {
		'totalSpend': total_value,
		'monthlySpend': monthly,
		'topExpensiveItems': top_items,
		'categoryBreakdown': category_breakdown,
		'usageTrends': trends,
	}
	return forecasts, analytics

